http/binding/binding_test.go

```diff
@@ -0,0 +1,472 @@
+	String []string `form:"Str,omitempty" json:"string" binding:"required,dive,printascii"`
+	audit.ChangeRequest
+}
+
+func testJSONHandler(c *gin.Context) {
+	var t testStruct
+	if err := binding.BindJSON(c, ctxKey, &t); err != nil {
+		c.AbortWithStatus(errors.Code(err))
+	}
+	c.JSON(http.StatusOK, t)
+}
+
+func testQueryHandler(c *gin.Context) {
+	var t testStruct
+	if err := binding.BindQuery(c, ctxKey, &t); err != nil {
+		c.AbortWithStatusJSON(errors.Code(err), err)
+	}
+	c.JSON(http.StatusOK, t)
+}
+
+func testChangeRequestHandler(c *gin.Context) {
+	var t testChangeStruct
+	if err := binding.BindChangeRequest(c, ctxKey, &t); err != nil {
+		c.AbortWithStatus(errors.Code(err))
+	}
+	c.JSON(http.StatusOK, t)
+}
+
+func testIDtHandler(c *gin.Context) {
+	id, err := binding.BindID(c)
+	if err != nil {
+		c.AbortWithStatus(errors.Code(err))
+	}
+	c.JSON(http.StatusOK, id)
+}
+
+type BindingSuite struct {
+	suite.Suite
+	router *gin.Engine
+}
+
+// SetupTest runs before each Test
+func (s *BindingSuite) SetupTest() {
+	gin.SetMode(gin.TestMode)
+	s.router = gin.Default()
+}
+
+func TestHandlers(t *testing.T) {
+	suite.Run(t, new(BindingSuite))
+}
+
+func (s *BindingSuite) Test_BindJSON_FromBodyBindError() {
+	requestBody := `{
+	    "number": [
+	        0
+	    ],
+	    "string": [
+	        "🇺🇸",
+	    ]
+	}`
+	req, err := http.NewRequest(http.MethodPatch, testJSONEndpoint, strings.NewReader(requestBody))
+	s.NoError(err)
+
+	s.router.PATCH(testJSONEndpoint, testJSONHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusBadRequest, r.Code)
+}
+
+func (s *BindingSuite) Test_BindJSON_FromBody() {
+	requestBody := `{
+	    "number": [
+	        1,
+	        2,
+	        3
+	    ],
+	    "string": [
+	        "A",
+	        "B",
+	        "C"
+	    ]
+	}`
+	req, err := http.NewRequest(http.MethodPatch, testJSONEndpoint, strings.NewReader(requestBody))
+	s.NoError(err)
+
+	s.router.PATCH(testJSONEndpoint, testJSONHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusOK, r.Code)
+	s.Equal(r.Body.String(), "{\"number\":[1,2,3],\"string\":[\"A\",\"B\",\"C\"]}")
+}
+
+func (s *BindingSuite) Test_BindJSON_FromContext_OK() {
+	requestBody := `{
+	    "number": [
+	        1,
+	        2,
+	        3
+	    ],
+	    "string": [
+	        "A",
+	        "B",
+	        "C"
+	    ]
+	}`
+	req, err := http.NewRequest(http.MethodPatch, testJSONEndpoint, strings.NewReader(requestBody))
+	s.NoError(err)
+
+	testReq := &testStruct{
+		Number: []uint{1, 2, 3, 4},
+		String: []string{"A", "B", "C", "D"},
+	}
+	s.router.Use(func(c *gin.Context) {
+		c.Set(ctxKey, testReq)
+		c.Next()
+	})
+	s.router.PATCH(testJSONEndpoint, testJSONHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusOK, r.Code)
+
+	body, err := json.Marshal(&testReq)
+	s.NoError(err)
+	s.Equal(r.Body.String(), string(body))
```

Comment:
I think `body` is not what in the context

Reply:
This case is try to test we already have the request in context, the value are set here
https://github.com/wego/pkg/blob/a07d96900f5ed59ba1fe46be2810be24a5fe5435/http/binding/binding_test.go#L148-L151 @viettd

Reply:
Please ignore, my mistake 😂

Reply:
I changed the `testReq` to `inCtx`, make it easier to understand